code/headers/hardware_usart.hpp

```diff
@@ -5,151 +5,254 @@
+            Pio *const port<pioa> = PIOA;
+
+            template <>
+            Pio *const port<piob> = PIOB;
+
+            template <>
+            Pio *const port<piod> = PIOD;
+        } // namespace pio
+
+        namespace usart {
+            // helper structs for getting the pointer to the usart
+            template <typename USART>
+            Usart *const port = nullptr;
+
+            template <>
+            Usart *const port<usart0> = USART0;
+
+            template <>
+            Usart *const port<usart1> = USART1;
+
+            template <>
+            Usart *const port<usart2> = USART2;
+
+            template <>
+            Usart *const port<usart3> = USART3;
+        } // namespace usart
+
+        namespace periph_id {
+            // helper structs for getting the usart id
+            template <typename USART>
+            constexpr uint32_t id = 0;
+
+            template <>
+            constexpr uint32_t id<usart0> = ID_USART0;
+
+            template <>
+            constexpr uint32_t id<usart1> = ID_USART1;
+
+            template <>
+            constexpr uint32_t id<usart2> = ID_USART2;
+
+            template <>
+            constexpr uint32_t id<usart3> = ID_USART3;
+        } // namespace periph_id
+
+    } // namespace detail
+
+    template <typename Bus>
     class hardware_usart_c : public usart_connection_c {
     private:
-        Usart *hardware_usart = nullptr;
-        ringbuffer_c<uint8_t, BufferLength> input_buffer;
+        static inline ringbuffer_c<uint8_t, 256> input_buffer;
 
         /// @brief check if the transmitter is ready to send
         /// @return true if ready to send, false if not ready to send
-        bool transmit_ready() {
-            return (hardware_usart->US_CSR & 2);
-        }
-
-        /// @brief send a byte via usart
-        /// @param b: byte to send
-        void send_byte(const uint8_t &b) {
-            while (!transmit_ready()) {
-            }
-            hardware_usart->US_THR = b;
+        constexpr static bool transmit_ready() {
+            return (detail::usart::port<Bus>->US_CSR & 2);
         }
 
         /// @brief receive a byte by reading the US_RHR register
         /// @return byte received
-        uint8_t receive_byte() {
-            return hardware_usart->US_RHR;
+        constexpr static uint8_t receive_byte() {
+            return detail::usart::port<Bus>->US_RHR;
         }
 
-        void set_peripheral(Pio *pio, uint32_t mask, peripheral p) {
-            uint32_t t = pio->PIO_ABSR;
+        template <uint32_t mask>
+        constexpr void set_peripheral() {
+            uint32_t t = detail::pio::port<typename Bus::pio>->PIO_ABSR;
 
-            if (p == peripheral::peripheral_a) {
-                pio->PIO_ABSR &= (~mask & t);
+            if constexpr (std::is_same_v<typename Bus::periph,
+                                         detail::pio::periph_a>) {
+                detail::pio::port<typename Bus::pio>->PIO_ABSR = (~mask & t);
```

Comment:
Even in the 'library_only' `detail` namespace I'm really not comfortable with single letter variables.

Reply:
This was as in t of temp. What do you prefer?

Reply:
I'd prefer `temp` or a more descriptive name such as `AB_select` or something to that effect